metrics: Add extra clustermesh metrics

sayboras opened the pull request:
### Desciption

Please refer to commit messages for more details.

Fixes: #16393

### Testing

Testing was done for below cases:
- single cluster with kind locally
- clustermesh with kind clusters, thanks to Bruno for https://github.com/bmcustodio/kind-cilium-mesh, simple changes to support images from CI build
- remove one cluster to verify failure metrics

<details>
<summary>1 kind cluster</summary>

```
# no clustermesh metric available
root@kind-worker2:/home/cilium# cilium metrics list -p clustermesh  
Metric   Labels   Value

# To make sure that nothing breaks, and check node connectivity metrics
root@kind-worker2:/home/cilium# cilium metrics list -p node_connectivity
Metric                                     Labels                                                                                                                                                                                                                           Value
cilium_node_connectivity_latency_seconds   protocol="http" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="10.1.0.93" target_node_name="kind-control-plane" target_node_type="remote_intra_cluster" type="endpoint"   0.000245
cilium_node_connectivity_latency_seconds   protocol="http" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="10.1.1.214" target_node_name="kind-worker" target_node_type="remote_intra_cluster" type="endpoint"         0.000225
cilium_node_connectivity_latency_seconds   protocol="http" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="10.1.2.229" target_node_name="kind-worker2" target_node_type="local_node" type="endpoint"                  0.000218
cilium_node_connectivity_latency_seconds   protocol="http" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="172.18.0.5" target_node_name="kind-worker2" target_node_type="local_node" type="node"                      0.000144
cilium_node_connectivity_latency_seconds   type="node" protocol="http" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="172.18.0.7" target_node_name="kind-worker" target_node_type="remote_intra_cluster"             0.000243
cilium_node_connectivity_latency_seconds   source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="172.18.0.8" target_node_name="kind-control-plane" target_node_type="remote_intra_cluster" type="node" protocol="http" source_cluster="kind-kind"      0.000203
cilium_node_connectivity_latency_seconds   protocol="icmp" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="10.1.0.93" target_node_name="kind-control-plane" target_node_type="remote_intra_cluster" type="endpoint"   0.000468
cilium_node_connectivity_latency_seconds   source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="10.1.1.214" target_node_name="kind-worker" target_node_type="remote_intra_cluster" type="endpoint" protocol="icmp" source_cluster="kind-kind"         0.000473
cilium_node_connectivity_latency_seconds   source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="10.1.2.229" target_node_name="kind-worker2" target_node_type="local_node" type="endpoint" protocol="icmp" source_cluster="kind-kind"                  0.000445
cilium_node_connectivity_latency_seconds   target_node_ip="172.18.0.5" target_node_name="kind-worker2" target_node_type="local_node" type="node" protocol="icmp" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind"                      0.000483
cilium_node_connectivity_latency_seconds   source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="172.18.0.7" target_node_name="kind-worker" target_node_type="remote_intra_cluster" type="node" protocol="icmp" source_cluster="kind-kind"             0.000471
cilium_node_connectivity_latency_seconds   type="node" protocol="icmp" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_ip="172.18.0.8" target_node_name="kind-control-plane" target_node_type="remote_intra_cluster"      0.000449
cilium_node_connectivity_status            source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_name="kind-control-plane" target_node_type="remote_intra_cluster" type="endpoint"                                              1.000000
cilium_node_connectivity_status            target_node_name="kind-control-plane" target_node_type="remote_intra_cluster" type="node" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind"                                                  1.000000
cilium_node_connectivity_status            source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_name="kind-worker" target_node_type="remote_intra_cluster" type="endpoint"                                                     1.000000
cilium_node_connectivity_status            source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_name="kind-worker" target_node_type="remote_intra_cluster" type="node"                                                         1.000000
cilium_node_connectivity_status            target_node_type="local_node" type="endpoint" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_name="kind-worker2"                                                              1.000000
cilium_node_connectivity_status            target_node_type="local_node" type="node" source_cluster="kind-kind" source_node_name="kind-worker2" target_cluster="kind-kind" target_node_name="kind-worker2"                                                                  1.000000


```

</details>

<details>
<summary>2 kind clusters, each has 1 control plane and 2 worker nodes</summary>

```
root@kind-1-worker:/home/cilium# cilium metrics list -p clustermesh
Metric                                               Labels                                                                             Value
cilium_clustermesh_global_services                   source_cluster="kind-1" source_node_name="kind-1-worker"                           4.000000
cilium_clustermesh_remote_cluster_nodes              source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2"   3.000000
cilium_clustermesh_remote_cluster_readiness_status   target_cluster="kind-2" source_cluster="kind-1" source_node_name="kind-1-worker"   1.000000
cilium_clustermesh_remote_clusters                   source_node_name="kind-1-worker" source_cluster="kind-1"                           1.000000

root@kind-1-worker:/home/cilium# cilium metrics list -p node_connectivity
Metric                                     Labels                                                                                                                                                                                                                         Value
cilium_node_connectivity_latency_seconds   protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.0.145" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="endpoint"   0.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-1-worker" target_node_type="local_node" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.1.63"                     0.000224
cilium_node_connectivity_latency_seconds   target_cluster="kind-1" target_node_ip="10.1.3.100" target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker"         0.000303
cilium_node_connectivity_latency_seconds   source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.2" target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="node" protocol="http"             0.000313
cilium_node_connectivity_latency_seconds   type="node" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.4" target_node_name="kind-1-worker" target_node_type="local_node"                        0.000351
cilium_node_connectivity_latency_seconds   source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.6" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="node" protocol="http"       0.000000
cilium_node_connectivity_latency_seconds   source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.0.84" target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="endpoint" protocol="http" source_cluster="kind-1"    0.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.1.118"         0.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.2.3"            0.000000
cilium_node_connectivity_latency_seconds   type="node" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.7" target_node_name="kind-2-worker" target_node_type="remote_inter_cluster"              0.000433
cilium_node_connectivity_latency_seconds   protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.8" target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="node"       0.000357
cilium_node_connectivity_latency_seconds   source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.9" target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="node" protocol="http"             0.000441
cilium_node_connectivity_latency_seconds   protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.0.145" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="endpoint"   0.000508
cilium_node_connectivity_latency_seconds   target_cluster="kind-1" target_node_ip="10.1.1.63" target_node_name="kind-1-worker" target_node_type="local_node" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker"                     0.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.3.100"         0.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.2"             0.000525
cilium_node_connectivity_latency_seconds   target_cluster="kind-1" target_node_ip="172.18.0.4" target_node_name="kind-1-worker" target_node_type="local_node" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker"                        0.000520
cilium_node_connectivity_latency_seconds   target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.6"       0.000510
cilium_node_connectivity_latency_seconds   target_node_type="remote_inter_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.0.84" target_node_name="kind-2-control-plane"    0.000000
cilium_node_connectivity_latency_seconds   target_node_type="remote_inter_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.1.118" target_node_name="kind-2-worker2"         0.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.2.3"            0.000000
cilium_node_connectivity_latency_seconds   protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.7" target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="node"              0.000531
cilium_node_connectivity_latency_seconds   target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.8"       0.000525
cilium_node_connectivity_latency_seconds   target_node_ip="172.18.0.9" target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2"             0.000538
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="endpoint"                                               1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="node"                                                   0.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker" target_node_type="local_node" type="endpoint"                                                                1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker" target_node_type="local_node" type="node"                                                                    1.000000
cilium_node_connectivity_status            source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="endpoint" source_cluster="kind-1"                                                     1.000000
cilium_node_connectivity_status            target_node_type="remote_intra_cluster" type="node" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker2"                                                         1.000000
cilium_node_connectivity_status            target_node_type="remote_inter_cluster" type="endpoint" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-control-plane"                                               1.000000
cilium_node_connectivity_status            target_cluster="kind-2" target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="node" source_cluster="kind-1" source_node_name="kind-1-worker"                                                   1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="endpoint"                                                      1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="node"                                                          1.000000
cilium_node_connectivity_status            target_node_type="remote_inter_cluster" type="endpoint" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-worker2"                                                     1.000000
cilium_node_connectivity_status            target_cluster="kind-2" target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="node" source_cluster="kind-1" source_node_name="kind-1-worker"                                                         1.000000
root@kind-1-worker:/home/cilium# 
```

</details>


</details>

<details>
<summary>2 kind clusters, delete one cluster to verify failures</summary>

```
Metric                                               Labels                                                                              Value
cilium_clustermesh_global_services                   source_cluster="kind-1" source_node_name="kind-1-worker2"                           4.000000
cilium_clustermesh_remote_cluster_failures           source_cluster="kind-1" source_node_name="kind-1-worker2" target_cluster="kind-2"   1.000000
cilium_clustermesh_remote_cluster_last_failure_ts    source_cluster="kind-1" source_node_name="kind-1-worker2" target_cluster="kind-2"   1642078024.643625
cilium_clustermesh_remote_cluster_nodes              source_cluster="kind-1" source_node_name="kind-1-worker2" target_cluster="kind-2"   0.000000
cilium_clustermesh_remote_cluster_readiness_status   source_cluster="kind-1" source_node_name="kind-1-worker2" target_cluster="kind-2"   0.000000
cilium_clustermesh_remote_clusters                   source_node_name="kind-1-worker2" source_cluster="kind-1"                           1.000000


Metric                                     Labels                                                                                                                                                                                                                         Value
cilium_node_connectivity_latency_seconds   target_node_type="remote_intra_cluster" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.0.86" target_node_name="kind-1-control-plane"    -1.000000
cilium_node_connectivity_latency_seconds   target_node_ip="10.1.1.63" target_node_name="kind-1-worker" target_node_type="local_node" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1"                     -1.000000
cilium_node_connectivity_latency_seconds   type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.3.100" target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster"         -1.000000
cilium_node_connectivity_latency_seconds   source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.2" target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="node" protocol="http"             -1.000000
cilium_node_connectivity_latency_seconds   source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.4" target_node_name="kind-1-worker" target_node_type="local_node" type="node" protocol="http"                        -1.000000
cilium_node_connectivity_latency_seconds   source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.6" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="node" protocol="http" source_cluster="kind-1"       -1.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.0.225"   -1.000000
cilium_node_connectivity_latency_seconds   target_node_type="remote_inter_cluster" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.1.154" target_node_name="kind-2-worker2"         -1.000000
cilium_node_connectivity_latency_seconds   target_node_type="remote_inter_cluster" type="endpoint" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.2.57" target_node_name="kind-2-worker"           -1.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="node" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.7"              -1.000000
cilium_node_connectivity_latency_seconds   target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="node" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.8"       -1.000000
cilium_node_connectivity_latency_seconds   target_node_ip="172.18.0.9" target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="node" protocol="http" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2"             -1.000000
cilium_node_connectivity_latency_seconds   source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.0.86" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1"    0.000532
cilium_node_connectivity_latency_seconds   target_node_ip="10.1.1.63" target_node_name="kind-1-worker" target_node_type="local_node" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1"                     0.000540
cilium_node_connectivity_latency_seconds   target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="10.1.3.100"         0.000558
cilium_node_connectivity_latency_seconds   target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.2"             0.000560
cilium_node_connectivity_latency_seconds   source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_ip="172.18.0.4" target_node_name="kind-1-worker" target_node_type="local_node" type="node" protocol="icmp"                        0.000583
cilium_node_connectivity_latency_seconds   target_node_ip="172.18.0.6" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1"       0.000552
cilium_node_connectivity_latency_seconds   target_node_ip="10.1.0.225" target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2"   0.000557
cilium_node_connectivity_latency_seconds   target_node_ip="10.1.1.154" target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="endpoint" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2"         0.000555
cilium_node_connectivity_latency_seconds   source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="10.1.2.57" target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="endpoint" protocol="icmp"           0.000561
cilium_node_connectivity_latency_seconds   target_cluster="kind-2" target_node_ip="172.18.0.7" target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker"              0.000555
cilium_node_connectivity_latency_seconds   target_node_type="remote_inter_cluster" type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.8" target_node_name="kind-2-control-plane"       0.000560
cilium_node_connectivity_latency_seconds   type="node" protocol="icmp" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_ip="172.18.0.9" target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster"             0.000555
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="endpoint"                                               1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-control-plane" target_node_type="remote_intra_cluster" type="node"                                                   0.000000
cilium_node_connectivity_status            target_node_type="local_node" type="endpoint" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker"                                                                1.000000
cilium_node_connectivity_status            type="node" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker" target_node_type="local_node"                                                                    0.000000
cilium_node_connectivity_status            type="endpoint" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster"                                                     1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-1" target_node_name="kind-1-worker2" target_node_type="remote_intra_cluster" type="node"                                                         0.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="endpoint"                                               1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-control-plane" target_node_type="remote_inter_cluster" type="node"                                                   0.000000
cilium_node_connectivity_status            source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-worker" target_node_type="remote_inter_cluster" type="endpoint" source_cluster="kind-1"                                                      1.000000
cilium_node_connectivity_status            target_node_type="remote_inter_cluster" type="node" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-worker"                                                          0.000000
cilium_node_connectivity_status            target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="endpoint" source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2"                                                     1.000000
cilium_node_connectivity_status            source_cluster="kind-1" source_node_name="kind-1-worker" target_cluster="kind-2" target_node_name="kind-2-worker2" target_node_type="remote_inter_cluster" type="node"                                                         0.000000

```

</details>

### TODO
- [x] Refactor Collect function for single responsibility, check if node connectivity metrics can be in status collector.
- [x] Update docs on metric page. Will do once the metrics are reviewed.
- [ ] ~~Update dashboard if making sense. Will do once the metrics are reviewed.~~
- [x] Capture metrics and run promtool for static check. Done as part of unit test for faster feedback.

```release-note
metrics: Add extra clustermesh metrics
```

sayboras commented on pkg/metrics/status.go:
```diff
@@ -154,4 +195,98 @@ func (s *statusCollector) Collect(ch chan<- prometheus.Metric) {
+		return defaults.ClusterName, nodeName
+	}
+	// remove forward slash at the end if any for cluster name
+	return path.Dir(clusterName), nodeName
```
this is kind of reversed logic of below function. I started with normal strings.Split based on slash separator, but not 100% sure if cluster name is having any slash.

https://github.com/cilium/cilium/blob/master/pkg/node/types/node.go#L200-L205

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -168,6 +168,34 @@ const (
+	LabelValueICMP     = "icmp"
+	LabelValueLocal    = "local"
+	LabelValueNode     = "node"
+	LabelValueRemote   = "remote"
```
These all have the prefix `LabelValue...`, however only some of them are alternatives for each other. I wonder if it would be easier to read & ensure that users correctly pick the right values by grouping alternatives with the same prefix, for example:

```suggestion
	LabelPeerEndpoint = "endpoint"
	LabelPeerNode     = "node"
	
	LabelTrafficHTTP     = "http"
	LabelTrafficICMP     = "icmp"
	
	LabelLocationLocal    = "local"
	LabelLocationRemote   = "remote"
```

joestringer commented on pkg/metrics/status.go:
```diff
@@ -154,4 +195,98 @@ func (s *statusCollector) Collect(ch chan<- prometheus.Metric) {
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityStatusDesc,
+			prometheus.GaugeValue,
+			boolToFloat64(isEndpointReachable),
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueEndpoint,
```
The IP address seems to be wrong here? Unless this is intentionally trying to group the metrics for both the endpoint + the node itself together without including the actual target IP for the health probe?

```suggestion
			localClusterName, localNodeName, targetClusterName, targetNodeName, endpointPathStatus.IP, nodeType, LabelValueEndpoint,
```

sayboras replied:
you are right, the IP here is wrong :100:

joestringer commented on pkg/metrics/status.go:
```diff
@@ -154,4 +195,98 @@ func (s *statusCollector) Collect(ch chan<- prometheus.Metric) {
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityLatencyDesc,
+			prometheus.GaugeValue,
+			metricValue,
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueEndpoint, LabelValueHTTP,
```
Similar question here as above whether the IP should correspond to the health probe that was performed here (ie the endpoint IP) or the node IP?

sayboras replied:
you are right, the IP here is wrong, it should be endpoint ip as the last label is endpoint (e.g. label peer endpoint)

joestringer commented on pkg/metrics/status.go:
```diff
@@ -154,4 +195,98 @@ func (s *statusCollector) Collect(ch chan<- prometheus.Metric) {
+		ch <- prometheus.MustNewConstMetric(s.nodeConnectivityLatencyDesc,
+			prometheus.GaugeValue,
+			metricValue,
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueEndpoint, LabelValueICMP,
```
Same question here also around `nodePathStatus.IP` vs. `endpointPathStatus.IP`.

joestringer commented on pkg/metrics/status.go:
```diff
@@ -154,4 +195,98 @@ func (s *statusCollector) Collect(ch chan<- prometheus.Metric) {
+			metricValue,
+			localClusterName, localNodeName, targetClusterName, targetNodeName, nodePathStatus.IP, nodeType, LabelValueNode, LabelValueICMP,
+		)
+	}
```
CC @gandro I know that you recently looked into the `PrimaryAddress` vs. `SecondaryAddresses` of health probes for dual-stack cases, do you have any thoughts on how we export these statistics through Prometheus?

gandro replied:
That is a very good question, but I don't have a good answer. 

Overall, I do think we *should report on the state of the secondary address*, as any errors on the secondary addresses will likely manifest as connectivity problems with IPv6 in dual-stack setups. The downstream metrics likely didn't consider them because the secondary address status was not exposed in the API until recently.

For the `status` (i.e. reachable vs non-reachable) I think we can consider the secondary paths easily by summarizing the status of all paths. Meaning: If one of the paths (e.g. secondary address via ICMP) to the node or endpoint is unhealthy, we report the whole target (node or endpoint) as unhealthy. This is what we already do for the existing metrics and brief output (see code below), and I think is consistent with the usage of `GetPathConnectivityStatusType` in the downstream community code which also summarizes HTTP and ICMP together.  After all, for full connectivity we need all paths to be healthy.

https://github.com/cilium/cilium/blob/8e25d64386426ea97fa7b3a35f77d074ef4a7546/pkg/health/client/client.go#L316-L317
https://github.com/cilium/cilium/blob/88d16c36037fb50e6838d849ff930ea0451ac711/pkg/metrics/status.go#L130-L144

For `latency` however, the story is more complicated. Here I think it's mathematically wrong to try to summarize and take the average of the latency of all paths, since each paths will have its own latency characteristics/distribution and therefore the average will pretty much be meaningless.

But by collecting the latency for each path separately will double our metrics, as we will have 8 latency values (`(node/endpoint) x (primary/secondary) x (icmp/http)` for each node to node pair instead of the existing 4.

sayboras replied:
Thanks both for your inputs. 

I can make update to aggregate status and expose value accordingly.

For latency, further to Sebastian's comment, SecondaryAddresses is a slice, so the total number of metrics per latency could potentially be even more. We can add feature gate like what we have right now (i.e. EnableNodeConnectivityLatencyMetric). Keen to hear your input here.

gandro replied:
The way `SecondaryAddresses` is implemented right now, it will contain none or at most one entry. I don't expect to change it any time soon, but good to keep in mind nonetheless.

Yeah, I would love for the metrics to be more configurable. For the Hubble metrics, one can configure each metric separately: https://docs.cilium.io/en/v1.11/operations/metrics/#hubble-metrics

Might be worth evaluating if that approach could also work for the normal Cilium metrics, instead of adding a feature flag for each metric.

sayboras replied:
The mechanism is a little bit different in cilium metrics, but functionality is there as long as we have these two new metrics in default list.

https://github.com/cilium/cilium/blob/master/Documentation/cmdref/cilium-agent.md
```
      --metrics strings                                      Metrics that should be enabled or disabled from the default metric list. (+metric_foo to enable metric_foo , -metric_bar to disable metric_bar)
```

gandro replied:
So in Hubble, I think the mechanism is even a bit more powerful: You can set options per metrics via this flag too. This would allow us to e.g. have something like `--metrics +health_status:include_secondary,+health_latency:include_secondary` to e.g. enabling the secondary addresses for these metrics. But I have not yet looked at how these configuration flags are parsed.

sayboras replied:
thanks, I can spin off the above metric configuration to another github issue if you don't mind.

gandro replied:
Sounds good to me!

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -168,6 +168,34 @@ const (
+	LabelTargetNodeName = "target_node_name"
+
+	// LabelTargetNodeType is the label for target node type (local vs remote)
+	LabelTargetNodeType = "target_node_type"
```
I think that this label is a bit ambiguous. Each agent will report statistics for:
1. Connectivity to itself (typically considered as `local`, eg in `cilium status` / `cilium-health status` output)
2. Connectivity to remote nodes in the same cluster
3. Connectivity to remote nodes in a different cluster

I would expect that on average, the latency would be in the order 1 << 2 < 3, and also it would be useful to understand that if connectivity is an issue, whether the problems are correlated with (2) vs. (3).

Which of (1), (2), (3) are intended to be handled here? Should we consider these as three separate values for the same label in prometheus, eg `local_node`, `remote_intra_cluster`, `remote_inter_cluster`?

joestringer commented on pkg/metrics/clustermesh_test.go:
```diff
@@ -0,0 +1,104 @@
+	})
+
+	// perform static checks such as prometheus naming convention, number of labels matching, etc
+	lintProblems, err := testutil.CollectAndLint(collector)
```
:tada: I didn't know about `CollectAndLint()`. Is this the same as running `promtool lint`?

sayboras replied:
Yes, the underlying rule is the same. Using promtool lint is ok, but a little slow in feedback loop, and it also requires actual metric output.

joestringer replied:
Ah neat yeah. In general I have been wondering how we could make the promtool linting of all of the metrics apply automatically, because most people who add new metrics are not aware of it and I didn't necessarily know the best way to recommend how to run the linter before.

We can certainly recommend to always add unit tests for each new metric and use `CollectAndLint()` to verify them, but unfortunately if developers aren't aware of this then they can easily skip the step and end up adding metrics which do not conform to the linter. Broadly one mitigation for this is we just make sure everyone in @cilium/metrics checks for this during review. If we could automatically lint all metrics in the daemon somehow then that would also be cool :thinking:

joestringer commented on pkg/metrics/clustermesh.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+func newClustermeshCollector() *clustermeshCollector {
+	ciliumClient, err := clientPkg.NewClient("")
```
All of the information that is being gathered here should already be present in the exact same running binary as where this is coming from, but the code is currently querying the agent's own API in order to gather statistics.

I wonder whether we could structure this a little differently, to just directly push the metrics from the code that tracks or generates these metrics in the first place. For instance, `c.totalRemoteClustersDesc` is generated from the Cilium API clustermesh -> clusters logic, which lives under `pkg/clustermesh`.

There are other examples in the tree of this sort of approach, like in `pkg/ipam/metrics/metrics.go`.

joestringer commented on pkg/metrics/clustermesh.go:
```diff
@@ -0,0 +1,238 @@
+var (
+	// sample remote cluster etcd status
+	// etcd: 1/1 connected, lease-ID=7c027e1caebe4533, lock lease-ID=7c027e1caebe4535, has-quorum=true: https://kind-cilium-mesh-1.mesh.cilium.io:31484 - 3.4.13 (Leader)
+	clusterMeshRemoteClusterEtcdStatusRegex = regexp.MustCompile(`^etcd: \d+/\d+ connected, lease-ID=([0-9a-f]+), lock lease-ID=([0-9a-f]+), has-quorum=true`)
```
If some portion of this code is derived from a similar implementation elsewhere such as https://github.com/form3tech-oss/extra-cilium-metrics/blob/master/main.go, I would suggest locating the original authors via `git blame` and attributing them also in the commit messages for any commits that include their code, for instance by adding `Co-Authored-By: John Doe <john@doe.org>`.

sayboras replied:
I personally don't like grepping details from free-style string, it's error prone, also others might update orignal status message to something else later. I am thinking about to expose details in more structured way in related code below. This is kind of aligned with alternative approach of exposing metrics _in line_.

https://github.com/cilium/cilium/blob/a8e3fa283cf40b330035265bf83d63800b296bcf/pkg/kvstore/etcd.go#L1153

joestringer replied:
:+1: yeah agree, seeing a random regex somewhere in the code always raises some questions ;-)

sayboras replied:
Based on https://github.com/cilium/cilium/pull/18313, there will be changes to remove consul and kvstore abstraction.

Currently, etcd related metrics are also abstracted, however, a couple of new metrics are kind of etcd specific, I reckon we might need to wait a little bit for the above PR to land in first.

joestringer replied:
I think we should try to avoid messing with the metrics as part of that work; if some metrics are already abstracted, my vote would be to keep them the same. If new metrics really are specific to etcd, I'm OK with adding them with an etcd-specific name. Or if we think that they should apply for any kvstore, we could abstract the name to say 'kvstore' rather than 'etcd'.

joestringer reviewed (changes requested):
Awesome to see some of the metrics that community users have valued becoming first-class metrics exposed directly by Cilium itself, thanks for picking this up :-) .

I've looked through, mostly at the first commit so far. I've pointed out a couple of minor nits here and there, but the biggest question I have is whether we can avoid querying the Cilium API for each prometheus scrape/collect request. Both of the sets of data that are being exposed here - cluster connectivity and clustermesh health - are now managed by the cilium-agent daemon itself, which is the one hosting the prometheus API. So the code shouldn't need to form REST queries to the same daemon to be able to reach over and read some memory elsewhere in the daemon in order to expose such metrics. Instead, I think it should be possible to just directly manage & update the prometheus objects from the subsystem directly (eg `pkg/clustermesh` or `pkg/health/server`). Have you considered this option?

sayboras committed:
metrics: Add connectivity metric inline

This commit is to add two new metrics (status and latency) between
any pair of cilium nodes.

Relates #16393

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

sayboras committed:
metrics: Add unit test for status collector metrics

This commit is to refactor status collector and add unit test for
the same.

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

sayboras committed:
metrics: Add clustermesh related metrics

This commit is to add below metrics for clustermesh:

- last failure timestamp per remote cluster
- readiness status per remote cluster
- total number of failures per remote cluster
- total number of nodes per remote cluster
- total number of global services
- total number of remote clusters

Relates #16393

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

joestringer commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -22,15 +25,20 @@ const (
 	configNotificationsChannelSize = 512
+
+	// subsystemClusterMesh is the subsystem for metrics
+	subsystemClusterMesh = "clustermesh"
```
nit: Usually we just name this `subsystem` in each subsystem, it's clear that this is specific to clustermesh since it's a private variable in this package.

joestringer commented on pkg/clustermesh/services.go:
```diff
@@ -31,14 +33,31 @@ func newGlobalService() *globalService {
+		metricTotalGlobalServices: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: metrics.Namespace,
+			Subsystem: subsystemClusterMesh,
+			Name:      "global_services",
```
Should this be `global_services_total`?

https://prometheus.io/docs/practices/naming/

(Similar for anything else that tracks a total count)

sayboras replied:
seems like both are passing promtool lint, but for the consistency, let me update it to global_services_total.

sayboras replied:
actually, below is the reason that I removed _total at the end.

```
... obtained []promlint.Problem = Text:"non-counter metrics should not have \"_total\" suffix"
```

joestringer replied:
:bulb: Weird, that seems like it contradicts the best practices from their own website on the link above :see_no_evil: 

Am I missing something? Is it because Gauges shouldn't represent accumulating counts?

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -1278,7 +1314,44 @@ func CreateConfiguration(metricsEnabled []string) (Configuration, []prometheus.C
+			NodeConnectivityStatus = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Name:      "node_connectivity_status",
+				Help:      "The last observed status of the connectivity between the current Cilium agent and other Cilium agents",
```
nit: It's from the current Cilium agent and other Cilium nodes - as these tests go to two locations: The node itself, and the simulated `cilium_health` endpoint running on that node (which runs in its own network namespace), and it tests both ICMP (ie only connectivity to the node) as well as HTTP (ie actual application connectivity).

joestringer commented on pkg/health/server/server.go:
```diff
@@ -170,9 +174,114 @@ func (s *Server) updateCluster(report *healthReport) {
+
+		// ICMP node secondary
+		for _, secondary := range nodePathSecondaryAddress {
+			collectConnectivityMetric(secondary.HTTP, localClusterName, localNodeName,
```
I was going to say "The comments about `ICMP node primary`, `ICMP node secondary` are not necessary, because the code describes exactly what it is doing." But then I noticed that the comment says "ICMP" but the metric says "HTTP". So I guess, the comments did help the review.

This looks like a bug, and also looks like the testing doesn't pick up on a problem like this.

sayboras replied:
Good spot. I have updated the latency values in test data so that all metrics are having different values, hopefully the test will help to pick up similar issue.

joestringer reviewed (changes requested):
Mostly minor nits, one bug below.

I think the only other reservation that I have is that I've heard that excessive labeling of metrics can be costly for performance of the metrics system. I do wonder whether these metrics reach a scalability limit when dealing with large clusters. As long as we have mitigations (such as being able to disable these metrics), this is probably fine for now and we'll gather feedback from users as they start to deploy these metrics in larger environments.

gandro reviewed (approved):
Awesome! I'm not very familiar with the internals of clustermesh, but the health part looks good to me.

ldelossa reviewed (approved):
approved contingent to other review feedback being addressed.
